catalog, catalogkv: add and use descriptor builder

postamar opened the pull request:
catalog,catalogkv: add and use descriptor builder
    
    Release justification: bug fixes and low-risk updates to new
    functionality.
    
    This commit is, at heart, a refactor of how catalog.Descriptor objects
    are created; either when read from storage via the catalogkv API
    or when constructed from an in-memory descpb.*Descriptor struct.
    
    The original motivation for this change was the fact that previously,
    calling the descriptor validation methods could sometimes modify the
    descriptor itself. Specifically, ValidateSelf would perform "privilege
    surgery" and modify the privileges descriptor. This appears to have been
    around since version 2 and is, in fact, responsible for a few bugs,
    which this commit also fixes.
    
    Taking a broader perspective it appears necessary that all such
    descriptor surgeries should be performed together when read from
    storage regardless of descriptor subtype (table, db, type, schema).
    These surgeries are typically no-ops: table format upgrade, foreign key
    representation upgrade, etc. Considering that catalog.Descriptor objects
    can be created either from storage or from memory, this involves:
      1. unifying the constructors in the (table|db|type|schema)desc
         packages,
      2. unifying the descriptor unwrapping logic in catalogkv.
    
    To address (1) this commit introduces catalog.DescriptorBuilder, a
    builder interface which replaces the NewImmutable, NewExistingMutable,
    NewCreatedMutable, etc. constructors. The builder does the same thing as
    these except it uses a deep copy of the proto descriptor instead of
    a shallow copy. This has, in itself, uncovered a few minor bugs.
    Still, the main reason for the existence of this builder object is that
    we can now perform all post-deserialization descriptor changes in a
    method to be called before building the catalog.Descriptor object
    proper.
    
    This commit also changes how we update the descriptor modification time
    using the MVCC timestamp. This is done at builder creation time, or, in
    cases where we stick to descpb structs, using a new
    descp.FromDescriptorWithMVCCTimestamp function which replaces
    descpb.TableFromDescriptor and the descpb.Descriptor.GetTable,
    GetDatabase, GetType and GetSchema methods. These methods are now all
    forbidden from casual use by a new linter check (it used to be just
    GetTable).
    
    To address (2) this commit uses the new builder to unify all descriptor
    wrapping logic in catalogkv into a new descriptorFromKeyValue function.
    This function is what backs the catalogkv API now. This commit also
    changes the API somewhat, the old GetDescriptorFromID function with its
    many flag arguments is gone in favour of many lightweight public getter
    functions (complementing existing ones) which all delegate to a common
    private getDescriptorByID function. The most important side benefit to
    doing all this is that we now no longer skip cross-reference validation
    when getting mutable descriptors through the catalogkv API.
    Cross-reference validation is now the default, and the exceptions to
    this are few:
      - The catalogkv DescGetter implemementation, aka the
        oneLevelUncachedDescGetter, does not perform these checks, because
        they, in turn, require a DescGetter. In theory we could get around
        this infinite recursion with caching but right now there's no point.
      - GetAllDescriptorsUnvalidated explicitly performs no validation at
        all.
    
    Doing all of this has uncovered a few bugs, the most serious of which is
    that the protected_ts_meta system table didn't have the correct
    permissions applied.
    
    Furthermore this commit also unifies namespace collision detection logic
    in catalogkv. A side-effect of this is that error messages for PG codes
    42710 and 42P07 contain qualified names more often than before.
    
    Release note (bug fix): Fixed privileges for system.protected_ts_meta.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/61429)
<!-- Reviewable:end -->

Commit:
catalog,catalogkv: add and use descriptor builder

Release justification: bug fixes and low-risk updates to new
functionality.

This commit is, at heart, a refactor of how catalog.Descriptor objects
are created; either when read from storage via the catalogkv API
or when constructed from an in-memory descpb.*Descriptor struct.

The original motivation for this change was the fact that previously,
calling the descriptor validation methods could sometimes modify the
descriptor itself. Specifically, ValidateSelf would perform "privilege
surgery" and modify the privileges descriptor. This appears to have been
around since version 2 and is, in fact, responsible for a few bugs,
which this commit also fixes.

Taking a broader perspective it appears necessary that all such
descriptor surgeries should be performed together when read from
storage regardless of descriptor subtype (table, db, type, schema).
These surgeries are typically no-ops: table format upgrade, foreign key
representation upgrade, etc. Considering that catalog.Descriptor objects
can be created either from storage or from memory, this involves:
  1. unifying the constructors in the (table|db|type|schema)desc
     packages,
  2. unifying the descriptor unwrapping logic in catalogkv.

To address (1) this commit introduces catalog.DescriptorBuilder, a
builder interface which replaces the NewImmutable, NewExistingMutable,
NewCreatedMutable, etc. constructors. The builder does the same thing as
these except it uses a deep copy of the proto descriptor instead of
a shallow copy. This has, in itself, uncovered a few minor bugs.
Still, the main reason for the existence of this builder object is that
we can now perform all post-deserialization descriptor changes in a
method to be called before building the catalog.Descriptor object
proper.

This commit also changes how we update the descriptor modification time
using the MVCC timestamp. This is done at builder creation time, or, in
cases where we stick to descpb structs, using a new
descp.FromDescriptorWithMVCCTimestamp function which replaces
descpb.TableFromDescriptor and the descpb.Descriptor.GetTable,
GetDatabase, GetType and GetSchema methods. These methods are now all
forbidden from casual use by a new linter check (it used to be just
GetTable).

To address (2) this commit uses the new builder to unify all descriptor
wrapping logic in catalogkv into a new descriptorFromKeyValue function.
This function is what backs the catalogkv API now. This commit also
changes the API somewhat, the old GetDescriptorFromID function with its
many flag arguments is gone in favour of many lightweight public getter
functions (complementing existing ones) which all delegate to a common
private getDescriptorByID function. The most important side benefit to
doing all this is that we now no longer skip cross-reference validation
when getting mutable descriptors through the catalogkv API.
Cross-reference validation is now the default, and the exceptions to
this are few:
  - The catalogkv DescGetter implemementation, aka the
    oneLevelUncachedDescGetter, does not perform these checks, because
    they, in turn, require a DescGetter. In theory we could get around
    this infinite recursion with caching but right now there's no point.
  - GetAllDescriptorsUnvalidated explicitly performs no validation at
    all.

Doing all of this has uncovered a few bugs, the most serious of which is
that the protected_ts_meta system table didn't have the correct
permissions applied.

Furthermore this commit also unifies namespace collision detection logic
in catalogkv. A side-effect of this is that error messages for PG codes
42710 and 42P07 contain qualified names more often than before.

Release note (bug fix): Fixed privileges for system.protected_ts_meta.

postamar commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -2364,9 +2364,9 @@ CREATE TABLE d.t1 (x d.farewell);
   AS OF SYSTEM TIME %s
 `, ts1))
-			sqlDB.ExpectErr(t, `pq: type "d.public.farewell" already exists`,
+			sqlDB.ExpectErr(t, `pq: type "farewell" already exists`,
```
This message change and others like it are caused by having unified the descriptor collision logic in `CheckObjectCollision` in `catalogkv`.

postamar commented on pkg/ccl/backupccl/backupbase/targets_test.go:
```diff
@@ -48,23 +49,22 @@ func TestDescriptorsMatchingTargets(t *testing.T) {
-			return desc
+		mkTable := func(descriptor tbDesc) catalog.Descriptor {
+			descProto := tabledesc.NewBuilder(&descriptor).BuildImmutable().DescriptorProto()
+			return catalogkv.NewBuilderWithMVCCTimestamp(descProto, ts1).BuildImmutable()
```
It's not necessary to use `NewBuilderWithMVCCTimestamp` here, strictly speaking, but it's the _nice_ way of maybe-setting a modification time.

postamar commented on pkg/ccl/backupccl/restore_job.go:
```diff
@@ -433,16 +433,7 @@ func WriteDescriptors(
-			descs = append(descs, db)
-		}
-		return catalog.ValidateSelfAndCrossReferences(ctx, bdg, descs...)
+		return nil
```
I looked into this in detail while chasing an unrelated bug and realized that there's no point to these checks here because the validation-on-write checks kick in very soon after.

postamar commented on pkg/ccl/backupccl/restore_job.go:
```diff
@@ -2075,30 +2066,21 @@ func getRestoringPrivileges(
-
-	var updatedPrivileges *descpb.PrivilegeDescriptor
-
+) (updatedPrivileges *descpb.PrivilegeDescriptor, err error) {
```
I believe this is a bug, we still have to update privileges on system tables when doing a cluster restore.

postamar commented on pkg/ccl/backupccl/show.go:
```diff
@@ -411,21 +408,24 @@ func nullIfEmpty(s string) tree.Datum {
-	var privDesc *descpb.PrivilegeDescriptor
+	b := catalogkv.NewBuilder(descriptor)
+	if b == nil {
+		return ""
```
This happens when `descriptor.Union == nil`.

postamar commented on pkg/ccl/backupccl/targets.go:
```diff
@@ -312,31 +309,6 @@ func lookupDatabaseID(
-	}
-	return nil
-}
-
```
I replaced this with `CheckObjectCollision` because it's the same thing.

postamar commented on pkg/ccl/importccl/import_processor_test.go:
```diff
@@ -914,15 +914,15 @@ func newTestSpec(
 	case roachpb.IOFileFormat_CSV:
 		descr = descForTable(ctx, t,
-			"CREATE TABLE simple (i INT PRIMARY KEY, s text )", 10, 20, NoFKs)
+			"CREATE TABLE simple (i INT PRIMARY KEY, s text )", 100, 200, NoFKs)
```
These descriptor IDs were in the system range and this caused privilege validation failures.

postamar commented on pkg/ccl/streamingccl/streamclient/random_stream_client.go:
```diff
@@ -223,7 +224,7 @@ func (m *randomStreamClient) getDescriptorAndNamespaceKVForTableID(
 		tableID,
 		fmt.Sprintf(RandomStreamSchemaPlaceholder, tableName),
-		&descpb.PrivilegeDescriptor{},
+		descpb.NewDefaultPrivilegeDescriptor(security.RootUserName()),
```
Now that we don't insta-upgrade our privilege descriptors when validating, we have to be more careful when crafting test cases.

postamar commented on pkg/sql/alter_database.go:
```diff
@@ -505,15 +514,6 @@ func (n *alterDatabasePrimaryRegionNode) switchPrimaryRegion(params runParams) e
-	if err != nil {
-		return err
-	}
-
```
This needed to be moved a tiny bit further up or else there would be cross-reference validation failures!

postamar commented on pkg/sql/alter_table_set_schema.go:
```diff
@@ -80,7 +80,7 @@ func (p *planner) AlterTableSetSchema(
 		if !dependent.ByID {
 			return nil, p.dependentViewError(
-				ctx, tableDesc.TypeName(), tableDesc.Name,
+				ctx, string(tableDesc.TypeName()), tableDesc.Name,
```
`TypeName()` now returns a `catalog.DescriptorType` which aliases `string`, for extra type safety.

postamar commented on pkg/sql/backfill.go:
```diff
@@ -802,7 +802,11 @@ func TruncateInterleavedIndexes(
 		// All the data chunks have been removed. Now also removed the
 		// zone configs for the dropped indexes, if any.
 		if err := db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
-			return RemoveIndexZoneConfigs(ctx, txn, execCfg, table.GetParentID(), indexes)
```
I'm certain that `table.GetParentID()` instead of `table.GetID()` is a bug here but found no way to test it. It only involves zone configs on interleaved indexes so I'm not super motivated to do more.

postamar commented on pkg/sql/backfill.go:
```diff
@@ -802,7 +802,11 @@ func TruncateInterleavedIndexes(
+			if err != nil {
+				return err
+			}
+			return RemoveIndexZoneConfigs(ctx, txn, execCfg, freshTableDesc, indexes)
```
I changed the signature of `RemoveIndexZoneConfigs` to take a `catalog.TableDescriptor` because otherwise it loads it from storage using a catalogkv api call which causes cross-reference validation check failures in `indexTruncateInTxn`.
This is another of these new failures which cropped up after I raised the validation level for mutable descs.

postamar commented on pkg/sql/set_zone_config.go:
```diff
@@ -949,21 +948,14 @@ func RemoveIndexZoneConfigs(
 		return nil
 	}
-	desc, err := catalogkv.GetDescriptorByID(ctx, txn, execCfg.Codec, tableID,
-		catalogkv.Mutable, catalogkv.TableDescriptorKind, true)
```
see https://github.com/cockroachdb/cockroach/pull/61429/files#r588774574

postamar commented on pkg/sql/catalog/descpb/privilege.go:
```diff
@@ -266,36 +266,40 @@ func MaybeFixPrivileges(id ID, ptr **PrivilegeDescriptor) bool {
-			return errors.AssertionFailedf("found no owner for system %s with ID=%d",
-				objectType, id)
+			return errors.AssertionFailedf("found no owner for %s%s with ID=%d",
+				maybeSystem, objectType, id)
```
For some reason the error messages here describe everything as "system" even when they're not system tables or anything.

postamar commented on pkg/sql/catalog/systemschema/system.go:
```diff
@@ -1423,12 +1428,12 @@ var (
 		NextIndexID: 2,
 		Privileges: descpb.NewCustomSuperuserPrivilegeDescriptor(
-			descpb.SystemAllowedPrivileges[keys.ReplicationStatsTableID], security.NodeUserName()),
+			descpb.SystemAllowedPrivileges[keys.ProtectedTimestampsMetaTableID], security.NodeUserName()),
```
bad copy pasta

postamar commented on pkg/sql/opt/optbuilder/scope.go:
```diff
@@ -35,7 +35,7 @@ import (
 // parent scope are also visible in this scope.
 //
-// See builder.go for more details.
+// See table_desc_builder.go for more details.
```
I'll fix this

postamar commented on pkg/sql/rename_table.go:
```diff
@@ -237,17 +234,6 @@ func (n *renameTableNode) startExec(params runParams) error {
-	} else if err != nil {
-		return err
-	}
-
```
This check is entirely redundant.

postamar commented on pkg/sql/repair.go:
```diff
@@ -569,7 +567,7 @@ func (p *planner) UnsafeDeleteDescriptor(ctx context.Context, descID int64, forc
 	var forceNoticeString string // for the event
 	if err != nil {
-		if !errors.Is(err, catalog.ErrDescriptorNotFound) && force {
+		if force {
```
I think `force` is enough here. The other test was failing because ErrDescriptorNotFound was being returned by a failing back-reference check.

postamar commented on pkg/sql/show_create.go:
```diff
@@ -256,6 +256,11 @@ func (p *planner) ShowCreate(
+		desc, err = lCtx.getTableByID(desc.GetID())
+		if err != nil {
+			return "", err
+		}
```
Previously this used to work without this because we were shallow-copying on descriptor construction. Deep-copy makes this necessary.

postamar commented on pkg/sql/tests/repair_test.go:
```diff
@@ -633,9 +633,9 @@ SELECT crdb_internal.unsafe_upsert_descriptor(59, crdb_internal.json_to_pb('cock
         { "privileges": 2, "user_proto": "newuser1" },
-        { "privileges": 2, "user_proto": "newuser2" }
+        { "privileges": 2, "user_proto": "newuser2" },
+        { "privileges": 2, "user_proto": "root" }
```
Apparently order matters here.

postamar commented on pkg/sql/tests/repair_test.go:
```diff
@@ -103,7 +103,7 @@ func TestDescriptorRepairOrphanedDescriptors(t *testing.T) {
 		require.NoError(t, crdb.ExecuteTx(ctx, db, nil, func(tx *gosql.Tx) error {
 			if _, err := tx.Exec(
-				"SELECT crdb_internal.unsafe_delete_descriptor($1);",
+				"SELECT crdb_internal.unsafe_delete_descriptor($1, true);",
```
Required due to more stringent validation-on-read.

postamar commented on pkg/sql/catalog/tabledesc/table_desc_builder.go:
```diff
@@ -0,0 +1,436 @@
+	dg catalog.DescGetter,
+	desc *descpb.TableDescriptor,
+	skipFKsWithNoMatchingTable bool,
+) (changes PostDeserializationTableDescriptorChanges, err error) {
```
moved as-is from structured.go, along with dependent functions

postamar commented on pkg/sql/catalog/catalogkv/desc_getter.go:
```diff
@@ -0,0 +1,91 @@
+// Copyright 2021 The Cockroach Authors.
```
moved all of this out of `catalogkv.go`, only big new thing is `fromKeyValue` method

ajwerner reviewed (commented):
> Reviewed 148 of 153 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61429)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @postamar)

___
*[pkg/ccl/backupccl/backup_test.go, line 2367 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MUu4B_1bJwuf0ph98_x-r1-2367:-MVKGWFuCTAhWa04XUTo:bm2sjf8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/ccl/backupccl/backup_test.go#L2367)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This message change and others like it are caused by having unified the descriptor collision logic in `CheckObjectCollision` in `catalogkv`.
</blockquote></details>

aye, not great though; now the error can be pretty ambiguous. 
___
*[pkg/ccl/backupccl/restore_job.go, line 436 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MUu4B_1bJwuf0ph98a1-r1-436:-MVKGhDv1ioz6Mtoh-3r:b-7m74ok) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/ccl/backupccl/restore_job.go#L436)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I looked into this in detail while chasing an unrelated bug and realized that there's no point to these checks here because the validation-on-write checks kick in very soon after.
</blockquote></details>

nice.
___
*[pkg/ccl/backupccl/restore_job.go, line 2069 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MUu4B_1bJwuf0ph98a1-r1-2069:-MVKGmcX6k3BCzZq53fz:b41xgw) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/ccl/backupccl/restore_job.go#L2069)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I believe this is a bug, we still have to update privileges on system tables when doing a cluster restore. 
</blockquote></details>

say more?
___
*[pkg/ccl/backupccl/targets.go, line 194 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MVKH4055b1W9Ad1V2kx:-MVKH4055b1W9Ad1V2ky:b-j3azfg) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/ccl/backupccl/targets.go#L194)):*
> ```Go
> 
> 				// Collect the prior IDs of table descriptors, as the ID may have been
> 				// changed during truncate.
> ```

can you add a ` prior to 20.2` after truncate. We no longer change the IDs.
___
*[pkg/ccl/changefeedccl/schemafeed/schema_feed.go, line 624 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MVKHrIaC6zZRRApdfd2:-MVKHrIaC6zZRRApdfd3:b4j20vx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/ccl/changefeedccl/schemafeed/schema_feed.go#L624)):*
> ```Go
> 
> 				b := catalogkv.NewBuilderWithMVCCTimestamp(&desc, k.Timestamp)
> 				if b != nil && (b.TypeName() == catalog.Table || b.TypeName() == catalog.Database) {
> ```

should this second one be `catalog.Type`?
___
*[pkg/sql/show_create.go, line 263 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MUzGRNurZ3K4XnvDvMF-r1-263:-MVKJX1i-QkCvjR7Ea-P:b-mlhmaw) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/sql/show_create.go#L263)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Previously this used to work without this because we were shallow-copying on descriptor construction. Deep-copy makes this necessary.
</blockquote></details>

oh man
___
*[pkg/sql/catalog/descriptor.go, line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MVK9mma4YpY7ihP_bTN:-MVK9mma4YpY7ihP_bTO:bcomsqg) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/sql/catalog/descriptor.go#L56)):*
> ```Go
> 	// TypeName returns a symbol identifying the type of the descriptor
> 	// built by this builder.
> 	TypeName() DescriptorType
> ```

nit: how about you call this method `DescriptorType()`? 
___
*[pkg/sql/catalog/descriptor.go, line 60 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MVK9w5sCUxll5uQCbfL:-MVK9w5sCUxll5uQCbfM:b-lqd6gp) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/sql/catalog/descriptor.go#L60)):*
> ```Go
> 	// RunPostDeserializationChanges attempts to perform post-deserialization
> 	// changes to the descriptor being built.
> 	RunPostDeserializationChanges(ctx context.Context, dg DescGetter) error
> ```

can you comment on the contract with the `DescGetter`?
___
*[pkg/sql/catalog/descriptor.go, line 109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MVKA7tV2KhuNdXPS9Qz:-MVKA7tV2KhuNdXPS9R-:b-bwbg6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/sql/catalog/descriptor.go#L109)):*
> ```Go
> 
> 	GetPrivileges() *descpb.PrivilegeDescriptor
> 	TypeName() DescriptorType
> ```

how about calling this `DescriptorType()` too?
___
*[pkg/sql/catalog/catalogkv/catalogkv.go, line 176 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MVKB2eR5AY9EiTNKTgV:-MVKB2eS6KVhJr9z2qOc:b-rhuq1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/sql/catalog/catalogkv/catalogkv.go#L176)):*
> ```Go
> 
> // builderFromKeyValue is a utility function for descriptorFromKeyValue which
> // unmarshals the proto  and checks that it exists and that it matches the
> ```

nit: extra space
___
*[pkg/sql/catalog/catalogkv/unwrap_validation_test.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MVKCBWL6VuN_hZWUlxV:-MVKCBWL6VuN_hZWUlxW:bfko3w) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/sql/catalog/catalogkv/unwrap_validation_test.go#L65)):*
> ```Go
> 		desc, err := m.GetDesc(context.Background(), id)
> 		require.NoErrorf(t, err, "id: %d", id)
> 		require.NotNil(t, desc, "id: %d", id)
> ```

`NotNilf`?
___
*[pkg/sql/catalog/descpb/descriptor.go, line 200 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MVKDeQy47ALidrS_yun:-MVKDeQz0L9tIQPgr6Wx:b-wkwnrt) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/sql/catalog/descpb/descriptor.go#L200)):*
> ```Go
> 
> // FromDescriptorWithMVCCTimestamp is a replacement for
> // Get(Table|Database|TypeSchema)() methods which seeks to ensure that clients
> ```

`Type|Schema`?
___
*[pkg/sql/catalog/descpb/privilege.go, line 302 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MUxqpi--2CN2pd8Ux8U-r1-302:-MVK8Sj1C67lw-TUVfxe:bi8gju) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/sql/catalog/descpb/privilege.go#L302)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

For some reason the error messages here describe everything as "system" even when they're not system tables or anything.
</blockquote></details>

I'm sure it's just code rot.
___
*[pkg/sql/catalog/descpb/privilege_test.go, line 353 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MVK8dynAYrbY2KmeZwf:-MVK8dynAYrbY2KmeZwg:bzf5nhn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/sql/catalog/descpb/privilege_test.go#L353)):*
> ```Go
> 
> 	rootWrongPrivilegesErr := "user root must have exactly SELECT, GRANT " +
> 		"privileges on s?y?s?t?e?m? ?table with ID=.*"
> ```

why not just `(system )?`
___
*[pkg/sql/catalog/systemschema/system.go, line 1431 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MUu4B_2qrQ9ULZQTkyh-r1-1431:-MVKENkyEJCN9kTi-QjV:b-fn8mpi) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/sql/catalog/systemschema/system.go#L1431)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

bad copy pasta
</blockquote></details>

:facepalm: 
___
*[pkg/sql/logictest/testdata/logic_test/enums, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MVKFHYI3SUa_nde2d7Q:-MVKFHYI3SUa_nde2d7R:b-s1ghdm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/sql/logictest/testdata/logic_test/enums#L28)):*
> ```
> SELECT * FROM db2.t
> 
> statement error pq: type "t" already exists
> ```

this highlights something interesting. Getting the error handling right here so that the right level understands enough context to construct a detailed error is not currently very easy.
___
*[pkg/sql/rowexec/tablereader.go, line 91 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MVKFixx4a2yZyMJIx0N:-MVKFixx4a2yZyMJIx0O:b-gjhiy1) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/sql/rowexec/tablereader.go#L91)):*
> ```Go
> 	tr.maxTimestampAge = time.Duration(spec.MaxTimestampAgeNanos)
> 
> 	tableDesc := tabledesc.NewBuilder(&spec.Table).BuildImmutableTable()
> ```

so, I feel like the copying behavior of the builder is going to hurt here. We end up calling this right here a *lot*. We should fix that so that instead we look up a cached descriptor but this new set of allocations per table reader most gives me pause to backporting this. 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 153 files at r1, 87 of 87 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-:-MVWey_K4Qg2wuDWZm4m:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt and @postamar)



<!-- Sent from Reviewable.io -->

postamar reviewed (commented):
Thanks for the review. Although we discussed things verbally I'll still respond to the comments here for the record.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-:-MVWmXKPBoNiZYvpyT2y:b3gtjaz)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @postamar)

___
*[pkg/ccl/backupccl/backup_test.go, line 2367 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MUu4B_1bJwuf0ph98_x-r1-2367:-MVWm_6g4iXpC2tJE176:bt6brli) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/ccl/backupccl/backup_test.go#L2367)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

aye, not great though; now the error can be pretty ambiguous. 
</blockquote></details>

Fixed by having `CheckObjectCollision` take a `tree.ObjectName` instead.
___
*[pkg/ccl/backupccl/restore_job.go, line 2069 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MUu4B_1bJwuf0ph98a1-r1-2069:-MVWnDh8DmdksB7xOX6m:b51cgms) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/ccl/backupccl/restore_job.go#L2069)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

say more?
</blockquote></details>

To clarify: the existing behaviour was correct _except_ when restoring to the temporary system DB. Those backed-up tables came from system tables and as such may not have r+w permissions for superusers, causing validation to freak out.

We never noticed this until now because r+w permissions for superusers used to be reset by the privilege descriptor surgery in ValidateSelf.

This change here grants those tables the same permissions as `restoreTempSystemDB` which is good enough, the contents of these restored tables get inserted into the real system tables later on.
___
*[pkg/ccl/backupccl/targets.go, line 194 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MVKH4055b1W9Ad1V2kx:-MVWqGu780OLJi6t0VXq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/ccl/backupccl/targets.go#L194)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

can you add a ` prior to 20.2` after truncate. We no longer change the IDs.
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/schemafeed/schema_feed.go, line 624 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MVKHrIaC6zZRRApdfd2:-MVWqIJj0r0DYUQKyUXe:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/ccl/changefeedccl/schemafeed/schema_feed.go#L624)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

should this second one be `catalog.Type`?
</blockquote></details>

Fixed.
___
*[pkg/sql/catalog/descriptor.go, line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MVK9mma4YpY7ihP_bTN:-MVWqKpn3CXo4anNol7A:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/sql/catalog/descriptor.go#L56)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: how about you call this method `DescriptorType()`? 
</blockquote></details>

Done.
___
*[pkg/sql/catalog/descriptor.go, line 60 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MVK9w5sCUxll5uQCbfL:-MVWqLyvCxARvgjFjCSy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/sql/catalog/descriptor.go#L60)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

can you comment on the contract with the `DescGetter`?
</blockquote></details>

Done.
___
*[pkg/sql/catalog/descriptor.go, line 109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MVKA7tV2KhuNdXPS9Qz:-MVWqMxo2CShiyrAGbqd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/sql/catalog/descriptor.go#L109)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

how about calling this `DescriptorType()` too?
</blockquote></details>

Done.
___
*[pkg/sql/catalog/catalogkv/catalogkv.go, line 176 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MVKB2eR5AY9EiTNKTgV:-MVWqOBC8BvJNW27Hu-6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/sql/catalog/catalogkv/catalogkv.go#L176)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: extra space
</blockquote></details>

Done.
___
*[pkg/sql/catalog/catalogkv/unwrap_validation_test.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MVKCBWL6VuN_hZWUlxV:-MVWqPueEQd5Tz5uVGb0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/sql/catalog/catalogkv/unwrap_validation_test.go#L65)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

`NotNilf`?
</blockquote></details>

Done.
___
*[pkg/sql/catalog/descpb/descriptor.go, line 200 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MVKDeQy47ALidrS_yun:-MVWqQs4EiB6lAO1PfN9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/sql/catalog/descpb/descriptor.go#L200)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

`Type|Schema`?
</blockquote></details>

Done.
___
*[pkg/sql/catalog/descpb/privilege_test.go, line 353 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MVK8dynAYrbY2KmeZwf:-MVWqTPs4lIcy4V8ymha:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/sql/catalog/descpb/privilege_test.go#L353)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

why not just `(system )?`
</blockquote></details>

Done
___
*[pkg/sql/rowexec/tablereader.go, line 91 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61429#-MVKFixx4a2yZyMJIx0N:-MVWqYJd6Aep5vJdd9Ah:b9j1oyy) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cbd4ba6f1099c95a536b7ab3561c1879f96714c/pkg/sql/rowexec/tablereader.go#L91)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

so, I feel like the copying behavior of the builder is going to hurt here. We end up calling this right here a *lot*. We should fix that so that instead we look up a cached descriptor but this new set of allocations per table reader most gives me pause to backporting this. 
</blockquote></details>

I ran the following microbenchmark: `BenchmarkKV/SQL` with the `kv` table modified to gratuitously fatten its descriptor. I did this by adding an extra 10 stored computed columns and an extra 10 secondary indexes. Using the `master` branch as a control I found the difference in allocs/op to not be significant. I was honestly expecting it to be worse than that. My working hypothesis right now is that the cost of these deep-copy allocations is amortized by the other allocations we already do when creating a tabledesc.immutable.


<!-- Sent from Reviewable.io -->

postamar commented:
bors r+

postamar commented:
rebased to fix merge conflict

postamar commented:
bors r+
